<launch>
  <arg name="maxDisappeared"         default="10" />
  <arg name="path_integ"         default="0.1" />
  <arg name="effect_x_range"         default="0.115" />
  <arg name="effect_y_range"         default="0.15" />
  <arg name="start_water_range"         default="0.15" />
  <arg name="finish_water_range"         default="0.07" />
  <arg name="path_x_max"         default="0.2" />
  <arg name="set_valve_pwm"         default="50" />
  <arg name="watering_mode"         default="3" />  
  <arg name="bias"         default="3" /> 
  <include file="$(find wapabawama_ros)/launch/detect.launch">
    <!-- <arg name="network_param_file" default="$(find darknet_ros)/config/lab_orchid_v4.yaml"/> -->
  </include>
  <include file="$(find wapabawama_ros)/launch/tf.launch"/>

  <!-- wapabawama_ros -->
  <group ns="wapabawama_ros">
    <include file="$(find wapabawama_ros)/launch/move.launch"/>
    <node name="SORT_left" pkg="wapabawama_ros" type="sort.py" output="screen">
        <param name="img_topic" value="/stereo/left/image_raw" />
        <param name="bbox" value="/left_camera/darknet_ros/bounding_boxes" />
        <param name="tracked_bbox" value="/left_camera/tracked_boxes" />
        <param name="window_name" value="left_window" />
    </node>
    <node name="SORT_right" pkg="wapabawama_ros" type="sort.py" output="screen">
        <param name="img_topic" value="/stereo/right/image_raw" />
        <param name="bbox" value="/right_camera/darknet_ros/bounding_boxes" />
        <param name="tracked_bbox" value="/right_camera/tracked_boxes" />
        <param name="window_name" value="right_window" />
    </node>
    <!-- Detect Lgv -->
    <node name="lgv_detector_left" pkg="wapabawama_ros" type="lgv_detector_test" output="screen">
      <param name="image" value="/stereo/left/image_raw" />
      <param name="bbox" value="/left_camera/tracked_boxes" />
      <param name="pub_lgvs_topic" value="/left/lgvs_tracked" />
      <param name="pub_img_topic" value="/left/pose_img" />
      <param name="alpha" value="236" />
      <param name="beta" value="1000" />
      <param name="cx" value="1530" />
      <param name="cy" value="425" />
      <param name="d" value="320" />
      <param name="box_expand" value="30" />
    </node>
    <node name="lgv_detector_right" pkg="wapabawama_ros" type="lgv_detector_test" output="screen">
      <param name="image" value="/stereo/right/image_raw" />
      <param name="bbox" value="/right_camera/tracked_boxes" />
      <param name="pub_lgvs_topic" value="/right/lgvs_tracked" />
      <param name="pub_img_topic" value="/right/pose_img" />
      <param name="alpha" value="237" />
      <param name="beta" value="1000" />  
      <param name="cx" value="164" />
      <param name="cy" value="385" />
      <param name="d" value="320" />
      <param name="biasx" value="0.02" />
      <param name="box_expand" value="30" />
    </node>
    <!-- Record Lgvs -->
    <node name="lgv_recorder" pkg="wapabawama_ros" type="lgv_recorder" output="screen">
      <param name="pub_lgvs_topic" value="lgvs_record" />
      <param name="left_record_limit" value="0.007788 " />
      <param name="right_record_limit" value="0.035787" />
    </node>
    <node name="follower_la1" pkg="wapabawama_ros" type="follower" output="screen">
      <param name="path_sub_name" value="la1/path"/>
      <param name="pose_pub_name" value="la1/set_pose"/>
      <param name="frame_name" value="la1"/>
      <param name="kp" value="0.05"/>
      <param name="integrator_max" value="0"/>
      <param name="orig" value="-0.14"/>
    </node>
    <node name="follower_la2" pkg="wapabawama_ros" type="follower" output="screen">
      <param name="path_sub_name" value="la2/path"/>
      <param name="pose_pub_name" value="la2/set_pose"/>
      <param name="frame_name" value="la2"/>
      <param name="kp" value="0.05"/>
      <param name="integrator_max" value="0"/>
      <param name="orig" value="0.10"/>
    </node>
    <node name="follower_la3" pkg="wapabawama_ros" type="follower" output="screen">
      <param name="path_sub_name" value="la3/path"/>
      <param name="pose_pub_name" value="la3/set_pose"/>
      <param name="frame_name" value="la3"/>
      <param name="kp" value="0.05"/>
      <param name="integrator_max" value="0"/>
      <param name="orig" value="-0.37"/>
    </node>
    <node name="path_planner_la1" pkg="wapabawama_ros" type="path_planner" output="screen">
      <param name="path_name"        value="la1/path"/>
      <param name="sub_topic"        value="lgvs_record"/>
      <param name="center_x"         value="0.01"/>
      <param name="max_y"            value="1"/>
      <param name="interval"         value="0.0001"/>
      <param name="lgv_dist_range"   value="$(arg effect_x_range)"/>
      <param name="lgv_effect_range" value="$(arg effect_y_range)"/>
      <param name="lgv_fusion_ratio" value="$(arg path_integ)"/>
      <param name="path_x_max"       value="$(arg path_x_max)"/>
    </node>
    <node name="path_planner_la2" pkg="wapabawama_ros" type="path_planner" output="screen">
      <param name="path_name"        value="la2/path"/>
      <param name="sub_topic"        value="lgvs_record"/>
      <param name="center_x"         value="0.25"/>
      <param name="max_y"            value="1"/>
      <param name="interval"         value="0.0001"/>
      <param name="lgv_dist_range"   value="$(arg effect_x_range)"/>
      <param name="lgv_effect_range" value="$(arg effect_y_range)"/>
      <param name="lgv_fusion_ratio" value="$(arg path_integ)"/>
      <param name="path_x_max"       value="$(arg path_x_max)"/>
    </node>
    <node name="path_planner_la3" pkg="wapabawama_ros" type="path_planner" output="screen">
      <param name="path_name"        value="la3/path"/>
      <param name="sub_topic"        value="lgvs_record"/>
      <param name="center_x"         value="-0.22"/>
      <param name="max_y"            value="1"/>
      <param name="interval"         value="0.0001"/>
      <param name="lgv_dist_range"   value="$(arg effect_x_range)"/>
      <param name="lgv_effect_range" value="$(arg effect_y_range)"/>
      <param name="lgv_fusion_ratio" value="$(arg path_integ)"/>
      <param name="path_x_max"       value="$(arg path_x_max)"/>
    </node>

    <node name="watering_vl1" pkg="wapabawama_ros" type="watering" output="screen">
      <param name="path_name"        value="la1/path"/>
      <param name="frame_name" value="la1"/>
      <param name="watering_mode" value="$(arg watering_mode)"/>
      <param name="pwm_name"         value="valve1/pwm"/>
      <param name="stop_sec"         value="2"/>
      <param name="bias"         value="0.04"/>   
      <param name="center_x"         value="0.01"/>
      <param name="set_pwm"            value="$(arg set_valve_pwm)"/>
      <param name="start_water_range" value="$(arg start_water_range)"/>
      <param name="finish_water_range" value="$(arg finish_water_range)"/>
    </node>
    <node name="watering_vl2" pkg="wapabawama_ros" type="watering" output="screen">
      <param name="path_name"        value="la2/path"/>
      <param name="frame_name" value="la2"/>
      <param name="watering_mode" value="$(arg watering_mode)"/>
      <param name="pwm_name"         value="valve2/pwm"/>
      <param name="stop_sec"         value="2"/>
      <param name="bias"         value="0.06"/>   
      <param name="center_x"         value="0.25"/>
      <param name="set_pwm"            value="$(arg set_valve_pwm)"/>
      <param name="start_water_range" value="$(arg start_water_range)"/>
      <param name="finish_water_range" value="$(arg finish_water_range)"/> 
    </node>
    <node name="watering_vl3" pkg="wapabawama_ros" type="watering" output="screen">
      <param name="path_name"        value="la3/path"/>
      <param name="frame_name" value="la3"/>
      <param name="watering_mode" value="$(arg watering_mode)"/>
      <param name="stop_sec"         value="2"/>
      <param name="bias"         value="0.04"/>   
      <param name="pwm_name"         value="valve3/pwm"/>
      <param name="center_x"         value="-0.22"/>
      <param name="set_pwm"            value="$(arg set_valve_pwm)"/>    
      <param name="start_water_range" value="$(arg start_water_range)"/>
      <param name="finish_water_range" value="$(arg finish_water_range)"/>
    </node>    

    <node name="record_1" pkg="wapabawama_ros" type="record_place">
      <param name="record_name"        value="la1/odem"/>
      <param name="frame_name"         value="la1"/>
    </node>
    <node name="record_2" pkg="wapabawama_ros" type="record_place">
      <param name="record_name"        value="la2/odem"/>
      <param name="frame_name"         value="la2"/>
    </node>
    <node name="record_3" pkg="wapabawama_ros" type="record_place">
      <param name="record_name"        value="la3/odem"/>
      <param name="frame_name"         value="la3"/>
    </node>

  </group>

  <node name="rviz" pkg="rviz" type="rviz" respawn="false" args="-d $(find wapabawama_ros)/resources/rviz2.rviz" />
</launch>
